src/duty/attestation/attestation.service.ts

```diff
@@ -76,48 +89,66 @@ export class AttestationService {
-      this.getCanonSlotRoot(attestation.target_epoch * this.slotsInEpoch),
-      this.getCanonSlotRoot(attestation.source_epoch * this.slotsInEpoch),
+      this.getCanonSlotRoot(attestation.targetEpoch * this.slotsInEpoch),
+      this.getCanonSlotRoot(attestation.sourceEpoch * this.slotsInEpoch),
     ]);
-    const attValidHead = attestation.head == canonHead;
-    const attValidTarget = attestation.target_root == canonTarget;
-    const attValidSource = attestation.source_root == canonSource;
-    const attIncDelay = Number(attestation.included_in_block - attestation.slot);
-    const isDencunFork = epoch >= this.dencunEpoch;
-    const flags = getFlags(attIncDelay, attValidSource, attValidTarget, attValidHead, isDencunFork);
-    for (const [valCommIndex, validatorIndex] of committee.entries()) {
-      const attHappened = attestation.bits.get(valCommIndex);
-      if (!attHappened) {
-        continue;
-      }
-      const processed = this.summary.epoch(attestation.target_epoch).get(validatorIndex);
-      if (!processed?.att_valid_source && flags.source) {
-        attestationFlags.source.push(validatorIndex);
+
+    const forkEpochs = await this.clClient.getForkEpochs();
+    const isDencunFork = epoch >= forkEpochs.dencun;
+    const isPectraFork = epoch >= forkEpochs.pectra;
+
+    const attValidHead = attestation.head === canonHead;
+    const attValidTarget = attestation.targetRoot === canonTarget;
+    const attValidSource = attestation.sourceRoot === canonSource;
+    const attIncDelay = Number(attestation.includedInBlock - attestation.slot);
+    const flags = getAttestationFlags(attIncDelay, attValidSource, attValidTarget, attValidHead, isDencunFork);
+
+    if (isPectraFork) {
+      let committeeOffset = 0;
+      for (const committeeIndex of attestation.committeeIndexes) {
+        // Each attestation corresponds to committee. Committee may have several aggregate attestations
+        const committee = committees.get(`${committeeIndex}_${attestation.slot}`);
+
+        if (committee == null) {
+          throw Error(`Committee ${committeeIndex} for attestation in slot ${attestation.slot} does not exist`);
+        }
+
+        for (const [valCommIndex, validatorIndex] of committee.entries()) {
+          const attHappened = attestation.aggregationBits.get(committeeOffset + valCommIndex);
+          if (!attHappened) {
+            continue;
+          }
+
+          this.processAttestationValidator(attestation, attIncDelay, validatorIndex, flags, attestationValidators);
+        }
+
+        committeeOffset += committee.length;
       }
-      if (!processed?.att_valid_target && flags.target) {
-        attestationFlags.target.push(validatorIndex);
+    } else {
+      const committee = committees.get(`${attestation.committeeIndex}_${attestation.slot}`);
+
+      if (committee == null) {
+        throw Error(`Committee ${attestation.committeeIndex} for attestation in slot ${attestation.slot} does not exist`);
       }
```

Comment:
here and after code the same as under `isPectraFork` except of `${attestation.committeeIndex}_${attestation.slot}`. can it be generalised somehow?

Reply:
OK, I agree. I moved this code to the common method.